File: FitBook.Model/Responses/Reservations/ReservationResponse.cs
using FitBook.Model.Enums;

namespace FitBook.Model.Responses.Reservations;

public class ReservationResponse : IEntityResponse
{
    public int Id { get; set; }

    public ReservationStatus Status { get; set; }

    public DateTime ReservedAtUtc { get; set; }
    public DateTime? ConfirmedAtUtc { get; set; }
    public DateTime? CancelledAtUtc { get; set; }
    public DateTime? CompletedAtUtc { get; set; }

    public string? CancellationReason { get; set; }

    public int UserAccountId { get; set; }
    public string UserFirstName { get; set; } = string.Empty;
    public string UserLastName { get; set; } = string.Empty;
    public string UserEmail { get; set; } = string.Empty;
    public int TrainingTermId { get; set; }

    public string TrainingName { get; set; } = string.Empty;
    public DateTime TrainingTermStartTimeUtc { get; set; }
    public DateTime TrainingTermEndTimeUtc { get; set; }

    public int? LastStatusChangedByUserAccountId { get; set; }

    public DateTime CreatedAtUtc { get; set; }
    public DateTime? UpdatedAtUtc { get; set; }
}

File: UI/fitbook_desktop/lib/models/responses/reservation_response.dart
import 'package:json_annotation/json_annotation.dart';

import '../enums/reservation_status.dart';

part 'reservation_response.g.dart';

/// Mirrors `FitBook.Model.Responses.ReservationResponse`.
@JsonSerializable()
class ReservationResponse {
  ReservationResponse({
    required this.id,
    required this.status,
    required this.reservedAtUtc,
    this.confirmedAtUtc,
    this.cancelledAtUtc,
    this.completedAtUtc,
    this.cancellationReason,
    required this.userAccountId,
    required this.userFirstName,
    required this.userLastName,
    required this.userEmail,
    required this.trainingTermId,
    required this.trainingName,
    required this.trainingTermStartTimeUtc,
    required this.trainingTermEndTimeUtc,
    this.lastStatusChangedByUserAccountId,
    required this.createdAtUtc,
    this.updatedAtUtc,
  });

  final int id;
  final ReservationStatus status;
  final DateTime reservedAtUtc;
  final DateTime? confirmedAtUtc;
  final DateTime? cancelledAtUtc;
  final DateTime? completedAtUtc;
  final String? cancellationReason;
  final int userAccountId;
  final String userFirstName;
  final String userLastName;
  final String userEmail;
  final int trainingTermId;
  final String trainingName;
  final DateTime trainingTermStartTimeUtc;
  final DateTime trainingTermEndTimeUtc;
  final int? lastStatusChangedByUserAccountId;
  final DateTime createdAtUtc;
  final DateTime? updatedAtUtc;

  factory ReservationResponse.fromJson(Map<String, dynamic> json) =>
      _$ReservationResponseFromJson(json);

  Map<String, dynamic> toJson() => _$ReservationResponseToJson(this);
}
